{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import sys\n",
    "import pandas as pd\n",
    "import glob\n",
    "import numpy as np\n",
    "\n",
    "import itertools\n",
    "from itertools import tee\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "virus_coord_dir = \"/Users/u241374/mike_tisza/sandbox/test_ct2_0802e/ct2_tmp/assess_prune/indiv_seqs\"\n",
    "\n",
    "gene_annotation_file = \"/Users/u241374/mike_tisza/sandbox/test_ct2_0802e/ct2_tmp/assess_prune/contig_gene_annotation_summary.tsv\"\n",
    "\n",
    "oriented_contigs_file = \"/Users/u241374/mike_tisza/sandbox/test_ct2_0802e/ct2_tmp/oriented_hallmark_contigs.fasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>chunk_start</th>\n",
       "      <th>chunk_stop</th>\n",
       "      <th>chunk_name</th>\n",
       "      <th>hallmark_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>test_ct2_0802e3952</td>\n",
       "      <td>0</td>\n",
       "      <td>12965</td>\n",
       "      <td>Chunk_0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>test_ct2_0802e4843</td>\n",
       "      <td>0</td>\n",
       "      <td>11550</td>\n",
       "      <td>Chunk_0</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>test_ct2_0802e4767</td>\n",
       "      <td>0</td>\n",
       "      <td>19541</td>\n",
       "      <td>Chunk_0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>test_ct2_0802e43597</td>\n",
       "      <td>0</td>\n",
       "      <td>62984</td>\n",
       "      <td>Chunk_0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>test_ct2_0802e4099</td>\n",
       "      <td>0</td>\n",
       "      <td>17549</td>\n",
       "      <td>Chunk_0</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                contig  chunk_start  chunk_stop chunk_name  hallmark_count\n",
       "0   test_ct2_0802e3952            0       12965    Chunk_0               6\n",
       "1   test_ct2_0802e4843            0       11550    Chunk_0              11\n",
       "2   test_ct2_0802e4767            0       19541    Chunk_0               4\n",
       "3  test_ct2_0802e43597            0       62984    Chunk_0               5\n",
       "4   test_ct2_0802e4099            0       17549    Chunk_0              10"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "try:\n",
    "\n",
    "    coord_files = glob.glob(os.path.join(virus_coord_dir, \"*.viruses.tsv\"))\n",
    "\n",
    "    df_from_each_coord = (pd.read_csv(coord, sep = \"\\t\", header = None,\n",
    "                                    names = [\"contig\", \"chunk_start\", \"chunk_stop\", \n",
    "                                             \"chunk_name\", \"hallmark_count\"])\n",
    "                                    for coord in coord_files)\n",
    "    chunk_coord_df = pd.concat(df_from_each_coord, ignore_index=True)\n",
    "\n",
    "    #chunk_coord_df.head()\n",
    "\n",
    "except:\n",
    "    print(\"no chunk coords\")\n",
    "\n",
    "chunk_coord_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_stop</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_orient</th>\n",
       "      <th>contig_length</th>\n",
       "      <th>dtr_seq</th>\n",
       "      <th>evidence_acession</th>\n",
       "      <th>evidence_description</th>\n",
       "      <th>Evidence_source</th>\n",
       "      <th>vscore_category</th>\n",
       "      <th>chunk_name</th>\n",
       "      <th>chunk_length</th>\n",
       "      <th>chunk_start</th>\n",
       "      <th>chunk_stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>test_ct2_0802e103</td>\n",
       "      <td>1</td>\n",
       "      <td>379</td>\n",
       "      <td>test_ct2_0802e103_9</td>\n",
       "      <td>+</td>\n",
       "      <td>6095</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CDD:COG4653</td>\n",
       "      <td>major-capsid-protein</td>\n",
       "      <td>hallmark_hmm</td>\n",
       "      <td>common_virus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>6095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>test_ct2_0802e103</td>\n",
       "      <td>432</td>\n",
       "      <td>999</td>\n",
       "      <td>test_ct2_0802e103_10</td>\n",
       "      <td>+</td>\n",
       "      <td>6095</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CDD:COG3740</td>\n",
       "      <td>head-maturation-protease</td>\n",
       "      <td>hallmark_hmm</td>\n",
       "      <td>common_virus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>6095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>test_ct2_0802e103</td>\n",
       "      <td>1000</td>\n",
       "      <td>2233</td>\n",
       "      <td>test_ct2_0802e103_11</td>\n",
       "      <td>+</td>\n",
       "      <td>6095</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CDD:COG4695</td>\n",
       "      <td>portal-protein</td>\n",
       "      <td>hallmark_hmm</td>\n",
       "      <td>common_virus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>6095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>test_ct2_0802e103</td>\n",
       "      <td>2216</td>\n",
       "      <td>2573</td>\n",
       "      <td>test_ct2_0802e103_12</td>\n",
       "      <td>+</td>\n",
       "      <td>6095</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDB:2KZ4_A</td>\n",
       "      <td>Putative-head-tail-adaptor</td>\n",
       "      <td>hallmark_hmm</td>\n",
       "      <td>common_virus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>6095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>test_ct2_0802e103</td>\n",
       "      <td>2559</td>\n",
       "      <td>2874</td>\n",
       "      <td>test_ct2_0802e103_13</td>\n",
       "      <td>+</td>\n",
       "      <td>6095</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CDD:TIGR01560</td>\n",
       "      <td>uncharacterized-phage-protein-(possible-DNA-pa...</td>\n",
       "      <td>hallmark_hmm</td>\n",
       "      <td>common_virus</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>6095</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              contig  gene_start  gene_stop             gene_name gene_orient  \\\n",
       "0  test_ct2_0802e103           1        379   test_ct2_0802e103_9           +   \n",
       "1  test_ct2_0802e103         432        999  test_ct2_0802e103_10           +   \n",
       "2  test_ct2_0802e103        1000       2233  test_ct2_0802e103_11           +   \n",
       "3  test_ct2_0802e103        2216       2573  test_ct2_0802e103_12           +   \n",
       "4  test_ct2_0802e103        2559       2874  test_ct2_0802e103_13           +   \n",
       "\n",
       "   contig_length  dtr_seq evidence_acession  \\\n",
       "0           6095      NaN       CDD:COG4653   \n",
       "1           6095      NaN       CDD:COG3740   \n",
       "2           6095      NaN       CDD:COG4695   \n",
       "3           6095      NaN        PDB:2KZ4_A   \n",
       "4           6095      NaN     CDD:TIGR01560   \n",
       "\n",
       "                                evidence_description Evidence_source  \\\n",
       "0                               major-capsid-protein    hallmark_hmm   \n",
       "1                           head-maturation-protease    hallmark_hmm   \n",
       "2                                     portal-protein    hallmark_hmm   \n",
       "3                         Putative-head-tail-adaptor    hallmark_hmm   \n",
       "4  uncharacterized-phage-protein-(possible-DNA-pa...    hallmark_hmm   \n",
       "\n",
       "  vscore_category chunk_name chunk_length  chunk_start  chunk_stop  \n",
       "0    common_virus         NA           NA            0        6095  \n",
       "1    common_virus         NA           NA            0        6095  \n",
       "2    common_virus         NA           NA            0        6095  \n",
       "3    common_virus         NA           NA            0        6095  \n",
       "4    common_virus         NA           NA            0        6095  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adjusted_gene_df = pd.DataFrame()\n",
    "\n",
    "gene_contig_df = pd.read_csv(gene_annotation_file, sep = \"\\t\")\n",
    "\n",
    "grouped_df = gene_contig_df.groupby('contig')\n",
    "\n",
    "for name, group in grouped_df:\n",
    "    contig_length1 = group['contig_length'].agg(pd.Series.mode)\n",
    "\n",
    "    dtr_status = group['dtr_seq'].agg(pd.Series.mode)\n",
    "\n",
    "    if len(dtr_status) == 0 and int(contig_length1.iloc[0]) >=10000:\n",
    "        ## lift genes from each chunk\n",
    "\n",
    "        ## @name is how to call the name variable in query\n",
    "        thisgroup_chunk_df = chunk_coord_df.query(\"contig == @name\")\n",
    "\n",
    "        for index, row in thisgroup_chunk_df.iterrows():\n",
    "            chunk_start = row.loc['chunk_start'] - 1\n",
    "            if chunk_start < 0:\n",
    "                chunk_start = 0\n",
    "            chunk_stop = row.loc['chunk_stop']\n",
    "            chunk_name = row.loc['chunk_name']\n",
    "            \n",
    "            thischunk_gene_df = group.copy().query(\"gene_start >= @chunk_start & gene_stop <= @chunk_stop\")\n",
    "\n",
    "            thischunk_gene_df['gene_start'] = thischunk_gene_df['gene_start'] - chunk_start\n",
    "            thischunk_gene_df['gene_stop'] = thischunk_gene_df['gene_stop'] - chunk_start\n",
    "            thischunk_gene_df['chunk_name'] = chunk_name\n",
    "            thischunk_gene_df['chunk_length'] = chunk_stop - chunk_start\n",
    "            thischunk_gene_df['chunk_start'] = chunk_start\n",
    "            thischunk_gene_df['chunk_stop'] = chunk_stop\n",
    "\n",
    "            adjusted_gene_df = pd.concat([adjusted_gene_df, thischunk_gene_df], ignore_index=True)\n",
    "\n",
    "    else:\n",
    "        #dtrs and seqs under 10kb\n",
    "        ## for these just return the gene coordinates as is\n",
    "        ## make chunk column\n",
    "\n",
    "        group['chunk_name'] = \"NA\"\n",
    "        group['chunk_length'] = \"NA\"\n",
    "        group['chunk_start'] = 0\n",
    "        group['chunk_stop'] = group['contig_length']\n",
    "\n",
    "        adjusted_gene_df = pd.concat([adjusted_gene_df, group], ignore_index=True)\n",
    "\n",
    "\n",
    "\n",
    "adjusted_gene_df.head()\n",
    "\n",
    "#adjusted_gene_df.to_csv(\"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>chunk_start</th>\n",
       "      <th>chunk_stop</th>\n",
       "      <th>new_chunk</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>test_ct2_0802e103</td>\n",
       "      <td>0</td>\n",
       "      <td>6095</td>\n",
       "      <td>test_ct2_0802e103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>test_ct2_0802e1136</td>\n",
       "      <td>0</td>\n",
       "      <td>3456</td>\n",
       "      <td>test_ct2_0802e1136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>test_ct2_0802e1184</td>\n",
       "      <td>0</td>\n",
       "      <td>1103</td>\n",
       "      <td>test_ct2_0802e1184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>test_ct2_0802e1537</td>\n",
       "      <td>0</td>\n",
       "      <td>1024</td>\n",
       "      <td>test_ct2_0802e1537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>test_ct2_0802e1610</td>\n",
       "      <td>0</td>\n",
       "      <td>1820</td>\n",
       "      <td>test_ct2_0802e1610</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                contig  chunk_start  chunk_stop           new_chunk\n",
       "0    test_ct2_0802e103            0        6095   test_ct2_0802e103\n",
       "10  test_ct2_0802e1136            0        3456  test_ct2_0802e1136\n",
       "14  test_ct2_0802e1184            0        1103  test_ct2_0802e1184\n",
       "16  test_ct2_0802e1537            0        1024  test_ct2_0802e1537\n",
       "18  test_ct2_0802e1610            0        1820  test_ct2_0802e1610"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_df = adjusted_gene_df[['contig', 'chunk_start', 'chunk_stop', 'chunk_name']].drop_duplicates()\n",
    "\n",
    "bed_df['new_chunk'] = np.where(bed_df['chunk_name'] == \"NA\", bed_df['contig'], bed_df['contig'] + \"@\" + bed_df['chunk_name'])\n",
    "\n",
    "bed_df = bed_df.drop('chunk_name', axis = 1)\n",
    "\n",
    "bed_df.head()\n",
    "#bed_df.to_csv(\"/Users/u241374/mike_tisza/sandbox/caccae_chunks.bed\", sep = \"\\t\", \n",
    "#              index = False, header = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ct2_dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
