{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "import subprocess\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## The following functions are copied wholesale from CheckV as I find them to be fast and correct\n",
    "## link https://bitbucket.org/berkeleylab/checkv/src/master/checkv/modules/complete_genomes.py\n",
    "\n",
    "def fetch_dtr(fullseq, min_length=20):\n",
    "    startseq = fullseq[0:min_length]\n",
    "    # find index positions of all matches of startseq in fullseq\n",
    "    # only keep matches occuring in 2nd half of string\n",
    "    matches = [\n",
    "        m.start() for m in re.finditer(\"(?={0})\".format(re.escape(startseq)), fullseq)\n",
    "    ]\n",
    "    matches = [_ for _ in matches if _ >= len(fullseq) / 2]\n",
    "    for matchpos in matches:\n",
    "        # determine if the match extends to the contig end\n",
    "        endseq = fullseq[matchpos:]\n",
    "        if fullseq[0 : len(endseq)] == endseq:\n",
    "            return endseq\n",
    "    return \"\"\n",
    "\n",
    "def reverse_complement(seq):\n",
    "    if sys.version_info > (3, 0):\n",
    "        trans = str.maketrans(\"ACTG\", \"TGAC\")\n",
    "    else:\n",
    "        trans = string.maketrans(\"ACTG\", \"TGAC\")\n",
    "    return seq[::-1].translate(trans)\n",
    "\n",
    "def fetch_itr(seq, min_len=20, max_len=1000):\n",
    "    rev = reverse_complement(seq)\n",
    "    # see if minimal substring occurs at end\n",
    "    if seq[:min_len] == rev[:min_len]:\n",
    "        # extend to maximum substring, up to <max_len>\n",
    "        i = min_len + 1\n",
    "        while seq[:i] == rev[:i] and i <= max_len:\n",
    "            i += 1\n",
    "        return seq[: i - 1]\n",
    "    # no match\n",
    "    else:\n",
    "        return \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                     contig  in_length_contig  out_length_contig dtr_seq  \\\n",
      "0                         1              6486               6484      NA   \n",
      "1                         2              4110               4108      NA   \n",
      "2                         3              4425               4423      NA   \n",
      "3                         4              3390               3388      NA   \n",
      "4                         5              3562               3560      NA   \n",
      "...                     ...               ...                ...     ...   \n",
      "6470  contig-100_1004.51897              1002               1000      NA   \n",
      "6471  contig-100_1005.51898              1002               1000      NA   \n",
      "6472  contig-100_1006.51899              1002               1000      NA   \n",
      "6473  contig-100_1007.51900              1001                999      NA   \n",
      "6474  contig-100_1008.51901              1000                998      NA   \n",
      "\n",
      "     itr_seq  \n",
      "0         NA  \n",
      "1         NA  \n",
      "2         NA  \n",
      "3         NA  \n",
      "4         NA  \n",
      "...      ...  \n",
      "6470      NA  \n",
      "6471      NA  \n",
      "6472      NA  \n",
      "6473      NA  \n",
      "6474      NA  \n",
      "\n",
      "[6475 rows x 5 columns]\n"
     ]
    }
   ],
   "source": [
    "fasta_file = \"/Users/u241374/mike_tisza/sandbox/test123_SRS893334.fna\"\n",
    "\n",
    "output_file = \"/Users/u241374/mike_tisza/sandbox/trimmed_circles1.fna\"\n",
    "\n",
    "if os.path.isfile(output_file):\n",
    "    os.remove(output_file)\n",
    "\n",
    "terminal_r_list = []\n",
    "for seq_record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "    if not len(seq_record.seq) >= 1000:\n",
    "        continue\n",
    "    dtr_seq = fetch_dtr(str(seq_record.seq))\n",
    "\n",
    "    if not dtr_seq:\n",
    "        dtr_seq = \"NA\"\n",
    "\n",
    "    itr_seq = fetch_itr(str(seq_record.seq))\n",
    "\n",
    "    if not itr_seq:\n",
    "        itr_seq = \"NA\"\n",
    "\n",
    "    if not dtr_seq == \"NA\":\n",
    "        print(f\">{seq_record.id}\", file = open(output_file, \"a\"))\n",
    "        print(seq_record.seq[:-len(dtr_seq)], file = open(output_file, \"a\"))\n",
    "\n",
    "        terminal_r_list.append([seq_record.id, len(seq_record.seq), len(seq_record.seq[:-len(dtr_seq)]), dtr_seq, itr_seq])\n",
    "    else:\n",
    "        print(f\">{seq_record.id}\", file = open(output_file, \"a\"))\n",
    "        print(seq_record.seq, file = open(output_file, \"a\"))\n",
    "\n",
    "        terminal_r_list.append([seq_record.id, len(seq_record.seq), len(seq_record.seq[:-len(dtr_seq)]), dtr_seq, itr_seq])\n",
    "\n",
    "terminal_df = pd.DataFrame(terminal_r_list, columns=[\"contig\", \"in_length_contig\", \"out_length_contig\", \"dtr_seq\", \"itr_seq\"])\n",
    "\n",
    "print(terminal_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ct2_dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
